src/ucp/tag/eager_rcv.c

```diff
@@ -74,22 +74,24 @@ ucp_eager_handler(void *arg, void *data, size_t length, void *desc,
         }
     }
 
+    rdesc = ucp_recv_desc_get(worker, data, desc, length, hdr_len, flags);
+    if (!rdesc) {
+        status = UCS_ERR_NO_MEMORY;
+        goto out;
+    }
     ucs_trace_req("unexp recv %c%c%c tag %"PRIx64" length %zu desc %p",
                   (flags & UCP_RECV_DESC_FLAG_FIRST) ? 'f' : '-',
                   (flags & UCP_RECV_DESC_FLAG_LAST)  ? 'l' : '-',
                   (flags & UCP_RECV_DESC_FLAG_EAGER) ? 'e' : '-',
                   recv_tag, length, rdesc);
 
-    if (data != rdesc + 1) {
-        memcpy(rdesc + 1, data, length);
-    }
-
-    rdesc->length  = length;
-    rdesc->hdr_len = hdr_len;
-    rdesc->flags   = flags;
     ucs_queue_push(&context->tag.unexpected, &rdesc->queue);
 
-    status = UCS_INPROGRESS;
+    if (flags & UCP_RECV_DESC_FLAG_SYNC) {
```

Comment:
Because this is not a direct callback in case of SYNC, return value is a part of protocol, see https://github.com/openucx/ucx/blob/master/src/ucp/tag/eager_rcv.c#L165. In the next PR, This place will be changed as well.